File: Day24.exs
# registers defined as [w, x, y, z]
defmodule Computer do
  def run(program, ipt) do
    run(program, [0, 0, 0, 0], ipt)
  end

  defp run([""], reg, _) do
    reg
  end

  defp run([curr | rest], reg, ipt) do
    {reg, ipt} = eval(String.split(curr, " "), reg, ipt)
    run(rest, reg, ipt)
  end

  defp eval(["inp", r], [w, x, y, z], [curr_ipt | rest]) do
    reg = case r do
      "w" -> [curr_ipt, x, y, z]
      "x" -> [w, curr_ipt, y, z]
      "y" -> [w, x, curr_ipt, z]
      "z" -> [w, x, y, curr_ipt]
    end
    {reg, rest}
  end

  defp eval(["add", a, b], reg, ipt) do
    res = arg(a, reg) + arg(b, reg)
    {updated_reg(a, res, reg), ipt}
  end

  defp eval(["mul", a, b], reg, ipt) do
    res = arg(a, reg) * arg(b, reg)
    {updated_reg(a, res, reg), ipt}
  end

  defp eval(["div", a, b], reg, ipt) do
    res = div(arg(a, reg), arg(b, reg))
    {updated_reg(a, res, reg), ipt}
  end

  defp eval(["mod", a, b], reg, ipt) do
    res = rem(arg(a, reg), arg(b, reg))
    {updated_reg(a, res, reg), ipt}
  end

  defp eval(["eql", a, b], reg, ipt) do
    res = if arg(a, reg) == arg(b, reg) do 1 else 0 end
    {updated_reg(a, res, reg), ipt}
  end

  defp arg(a, [w, x, y, z]) do
    case a do
      "w" -> w
      "x" -> x
      "y" -> y
      "z" -> z
      _ -> String.to_integer(a)
    end
  end

  defp updated_reg(key, val, reg) do
    case {key, reg} do
      {"w", [_, x, y, z]} -> [val, x, y, z]
      {"x", [w, _, y, z]} -> [w, val, y, z]
      {"y", [w, x, _, z]} -> [w, x, val, z]
      {"z", [w, x, y, _]} -> [w, x, y, val]
    end
  end
end


{:ok, ipt} = File.read("inputs/Day24.in")
prog = String.split(ipt, "\n")
#                              0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10
IO.inspect(Computer.run(prog, [1, 1, 1, 4, 1, 1, 9, 8, 1, 1, 1, 3, 1, 1]))
# ,  1,  9,  7,  9

# subprogs =
#   ipt
#   |> String.split("inp w\n")
#   |> tl()
#   |> Enum.map(
#     fn prog ->
#       subprog = String.split(prog, "\n")
#       ["inp w", "inp z" | subprog]
#     end)
# IO.inspect(subprogs)

# defmodule Solve do
#   def run([prog | subprogs], w_inputs \\ [], z_inputs \\ [0], memo \\ %{}) do
#     for i <- 9..1\\-1 do

#     end
#     memo
#   end

#   def run([""], w_inputs, z_inputs, memo) do
#     # Map.put(memo, {w_inputs, z}, z)
#     # z
#   end
# end

# [w, x, y, z] = Computer.run(prog, [9, 9, 6, 9, 9, 9, 9, 9, 2, 7])
# defmodule Solve do
#   def guess(prog, lst, [{upper, lower} | limit_tail], parent) do
#     for i <- (upper)..(lower)//-1 do
#       guess(prog, lst ++ [i], limit_tail, parent)
#     end
#     nil
#   end

#   def guess(prog, lst, [], parent) do
#     [_, _, _, z] = Computer.run(prog, lst)
#     z
#   end
# end

# for i <- 1..9 do
#   spawn_link(
#     Solve,
#     :guess,
#     [
#       prog,
#       [],
#       [
#         {1, 1},
#         {7, 7},
#         {6, 6},
#         {9, 9},
#         {9, 9},
#         {i, i},
#         {9, 1},
#         {9, 1},
#         {9, 1},
#         {9, 1},
#         {9, 1},
#         {9, 1},
#         {9, 1},
#         {9, 1}
#       ],
#       parent
#     ]
#   )
# end

# receive do
#   {:yo, {lst, z}} -> IO.inspect(lst, label: "z=#{z}")
# end




# IO.puts("#{w} #{x} #{y} #{z}")
# IO.puts("#{rem(z, 26) - 5}")
# Solve.guess(prog, [9, 9, 6, 9, 9, 9, 8])


# x = %{}
# x = Map.put(x, {1, 2}, 2)
# IO.inspect(x)

# l = fn lst -> if length(list) == 5 do IO.puts("done") else l() end end

File: Day24.py
from collections import deque

def arg(reg: dict[str, int], symbol: str):
    if symbol.isalpha():
        return reg[symbol]
    return int(symbol)


def run(prog: list[str], inp: list[int]):
    reg = {
        'w': 0,
        'x': 0,
        'y': 0,
        'z': 0
    }
    iptr = 0
    for line in prog:
        match line.split():
            case ['inp', r]:
                reg[r] = inp[iptr]
                iptr += 1
            case ['add', a, b]:
                reg[a] += arg(reg, b)
            case ['mul', a, b]:
                reg[a] *= arg(reg, b)
            case ['div', a, b]:
                reg[a] //= arg(reg, b)
            case ['mod', a, b]:
                reg[a] %= arg(reg, b)
            case ['eql', a, b]:
                reg[a] = 1 if reg[a] == arg(reg, b) else 0
    return reg


def main():
    with open('inputs/Day24.in') as f:
        block_strings = f.read().split('inp w\n')[1:]
        blocks = []
        for b in block_strings:
            blocks.append(['inp w', 'inp z'] + b.splitlines())
    w_checks = [
        None,
        None,
        None,
        lambda z: z % 26 - 5,
        None,
        None,
        None,
        lambda z: z % 26 - 14,
        lambda z: z % 26 - 8,
        None,
        lambda z: z % 26,
        lambda z: z % 26 - 5,
        lambda z: z % 26 - 9,
        lambda z: z % 26 - 1
    ]
    w = [9]
    z = [0]
    p = 0
    while True:
        if p == len(blocks) and z[-1] < 2000:
            if z[-1] == 0:
                print(w, z)
                break
            print(w, z[-1])
        while (w[p] < 1 or w[p] > 9) or (p == len(blocks) and z != 0):
            w.pop()
            z.pop()
            w[-1] -= 1
            p -= 1
        z.append(run(blocks[p], [w[p], z[p]])['z'])
        if w_checks[p]:
            w.append(w_checks[p](z[p]))
        else:
            w.append(9)
        p += 1

if __name__ == '__main__':
    main()
